model/fn-execution/src/main/proto/beam_fn_api.proto

```diff
@@ -199,6 +232,10 @@ message ProcessBundleResponse {
   // (Optional) If metrics reporting is supported by the SDK, this represents
   // the final metrics to record for this bundle.
   Metrics metrics = 1;
+
+  // (Optional) Specifies that the bundle has been split since the last
+  // ProcessBundleProgressResponse was sent.
+  BundleSplit split = 2;
```

Comment:
Why do we need this here or on the ProcessBundleProgressResponse?

It seems as though we should really only need it on ProcessBundleSplitResponse

Reply:
As answered on the document - I think the SDK should be able to split without being asked, that's why the split request is sort of "level-triggered" rather than edge-triggered, and the SDK can fulfill a split at any moment, possibly multiple times (e.g. it could see a deadline, and gradually return multiple residuals).

Reply:
I see, in that case it would be preferable if there was an SDK side initiated instruction that was "ShedBundleRequest" (as in its shedding work) and the runner could acknowledge the receipt or reject it which would tell the SDK that it needs to continue doing that work. This mechanism could be used to "checkpoint" since the primary bundle would be unset and the residual would be represented by a bunch of applications.

Unfortunately, API wise this will require some pretty substantive changes which does bring up a slew of clean-up tasks requested on the Fn API.

This would then mean that the ProcessBundleResponse/ProcessBundleProgressResponse would not contain BundleSplit.

Reply:
I agree that https://issues.apache.org/jira/browse/BEAM-3787 is the right way to go, however I'd like to make progress on the current PR and followup work and have some protos upon which to build the SDK harness and runner harness changes - I think as BEAM-3787 is implemented, it will be fairly easy to port onto that new API; I'm willing to do it for splitting in particular when it's time.